### PR TITLE
added Robot, which encaspulate frontend and backend

### PR DESCRIPTION
**Pull request by vincentberenz**
# What it is

The class Robot is a subclass of Frontend, which encapsulates a Backend.
This allows to construct a Frontend and a Backend in a single instance. This instance provides the same API as Frontend.

edit : In a supplementary commit, Robot takes the class RobotDriver as template, and construct the related instance. So, the frontend, the backend, the data and the driver instances are all instantiated via a single call.

# How it was tested

demo.cpp was extended to also use Robot

### CODE REVIEW

File: include/robot_interfaces/robot.hpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	DataPtr data_ptr = std::make_shared<Data>())
+    : RobotFrontend<Action,Observation>(data_ptr),
+    data_ptr_(data_ptr),
+    driver_ptr_(std::make_shared<Driver>(args ...)),
+    backend_(driver_ptr_,
+	     data_ptr_,
+	     max_action_duration_s,
+	     max_inter_action_duration_s)
+  {
+    // compile time checking template Driver inherate from RobotDriver
```
Comment (luator) [inline include/robot_interfaces/robot.hpp]: Typo "inherate"

File: include/robot_interfaces/robot.hpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ * @brief RobotFrontend that construct and encapsulates
+ * its related RobotBackend. It also construct and starts
+ * the robot driver.
+ */
+template<typename Action,
+	 typename Observation,
+	 typename Driver,
+	 typename ... Args>
```
Comment (luator) [inline include/robot_interfaces/robot.hpp]: It might be good extending the demo a bit to make use of this feature.

Comment (vincentberenz) [reply include/robot_interfaces/robot.hpp]: @luator 
Not sure what you mean, I also updated the demo to perform the two version (frontend/backend, and robot)

File: demos/demo.cpp [hunk 3]
```diff
@@ -145,42 +146,84 @@ int main()
...
+	action.print(false);
+	observation.print(true);
+      }
+  }
+
+  // demo representing usage of frontend and backend
+  // encapsulated in the same instance
+  {
+
+    std::cout << "\n -- * -- Robot -- * --\n" << std::endl;
```
Comment (vincentberenz) [inline demos/demo.cpp]: robot verson of the demo here

File: include/robot_interfaces/robot.hpp (hunk 2 above)
Comment (luator) [reply include/robot_interfaces/robot.hpp]: I mean the arguments for the driver (the `... Args` part).  In the demo, the driver does not take any arguments.  I am not familiar with this syntax, I assume it would work like the following?

    robot_interfaces::Robot<Action, Observation, Driver, int, double> robot(max_duration, max_inter, 42, 3.14);

Comment (vincentberenz) [reply include/robot_interfaces/robot.hpp]: Not intuitively, it would work as:

robot_interfaces::Robot<Action, Observation, Driver> robot(max_duration, max_inter, 42, 3.14);

+1 on updating the demo, will do

**Comment by luator** [comment]
What is the status here?

**Comment by vincentberenz** [comment]
I had to move to something else for a while, but this is still in progress. There are some bugs in the latest commit I found out and that I am correcting

File: demos/demo.cpp [hunk 4]
```diff
@@ -74,13 +75,14 @@ class Driver : public robot_interfaces::RobotDriver< Action,Observation>
   
 public:
 
-  Driver(){}
+  Driver(int min, int max)
```
Comment (vincentberenz) [inline demos/demo.cpp]: This shows an example of a Driver taking arguments

File: demos/demo.cpp [hunk 5]
```diff
@@ -145,42 +146,88 @@ int main()
...
+
+    std::cout << "\n -- * -- Robot -- * --\n" << std::endl;
+
+    typedef robot_interfaces::Robot<Action,Observation,Driver> Robot;
+
+    int min = 0;
+    int max = 100;
+    Robot robot(max_action_duration_s,
+		max_inter_action_duration_s,
+		min,
```
Comment (vincentberenz) [inline demos/demo.cpp]: Arguments for the driver are simply passed to the constructor of Robot

File: include/robot_interfaces/monitored_robot_driver.hpp [hunk 6]
```diff
@@ -172,32 +175,45 @@ class MonitoredRobotDriver : public RobotDriver<Action, Observation>
...
-            {
-                std::cout << "Action did not start on time, shutting down. Any "
-                             "further actions will be ignored."
-                          << std::endl;
-                shutdown();
-                return;
-            }
-        }
+	  {
+	    if(! std::isinf(max_action_duration_s_))
```
Comment (vincentberenz) [inline include/robot_interfaces/monitored_robot_driver.hpp]: This update is to make the class robust to infinite durations (without these changes it would hang indefinitely on hand). Note that this change is not related to the initial purpose of this pull request. Creating a new pull request sounds too much overhead to me, but will never do it if requested

File: include/robot_interfaces/robot.hpp (hunk 2 above)
Comment (vincentberenz) [reply include/robot_interfaces/robot.hpp]: done

**Comment by vincentberenz** [comment]
@luator 

"LGTM but please reformat the code before merging."

I ran the reformating (as implemented to mpi-cmake-modules), but this results in changes in the whole repo. What wuuld be the best way to proceed ?

File: include/robot_interfaces/monitored_robot_driver.hpp [hunk 7]
```diff
@@ -172,32 +172,45 @@ class MonitoredRobotDriver : public RobotDriver<Action, Observation>
...
+		      << "Action did not end on time, shutting down. Any further "
+		      "actions will be ignored."
+		      << std::endl;
+		    shutdown();
+		    return;
+		  }
+	      }
+	      else
+		{	    
+		  real_time_tools::Timer::sleep_ms(1);
```
[the remark this answers is not in the document]
Comment (vincentberenz) [reply include/robot_interfaces/monitored_robot_driver.hpp]: oops, you are right. will do

[the remark this answers is not in the document]
Comment (luator) [reply include/robot_interfaces/monitored_robot_driver.hpp]: Did you have time to look into this already?  I am actually running into the same issue right now (monitored driver hanging on destruction with infinite timeouts).

[the remark this answers is not in the document]
Comment (luator) [reply include/robot_interfaces/monitored_robot_driver.hpp]: I added a check in the monitored driver and simply don't start the thread if both timeouts are infinite ([see here](https://github.com/open-dynamic-robot-initiative/robot_interfaces/blob/d15cc3eb32396b80e801a1b3e611e290145c653f/include/robot_interfaces/monitored_robot_driver.hpp#L76)).  The corresponding PR will hopefully soon be merged, so you could simply wait for this.

However, if you already have a nicer solution, I'm happy to discard my fix and adopt yours :)

[the remark this answers is not in the document]
Comment (vincentberenz) [reply include/robot_interfaces/monitored_robot_driver.hpp]: done, I also applied clang_format

[the remark this answers is not in the document]
Comment (luator) [reply include/robot_interfaces/monitored_robot_driver.hpp]: Oh, I just saw, that you already pushed something regarding this.

**Comment by vincentberenz** [comment]
@luator merged !
